enhancements/user-workload-monitoring.md

```diff
@@ -0,0 +1,275 @@
+Other Red Hat products, outside the OpenShift framework components also
+don’t have a monitoring solution to integrate their metrics and out-of-the-box insights into their well being.
+Currently it is difficult to achieve metering/chargeback for Red Hat products without a central observability stack.
+
+### Goals
+
+Allow deployment and reconciliation of scrape targets, recording rules, and alerting rules in user namespaces.
+This effectively means the following custom resources:
+
+-  [ServiceMonitor](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#servicemonitor)
+- [PodMonitor](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#podmonitor)
+- [PrometheusRule](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#prometheusrule)
+
+Allow the correlation between app and cluster centric metrics.
+
+### Non-Goals
+
+Out of scope are:
+
+- Definition of Dashboards.
+- Deployment of Prometheus instances in user namespaces.
+- Deployment of Alertmanager instances in user namespaces.
+
+## Proposal
+
+![](user-monitoring-deployment.png)
+
+The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers
+and one additional Prometheus Operator (PO) reconciling user namespaces
+as well as the newly added Prometheus servers.
+
+The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.
+We will continue to use a centralized Alertmanager (A) cluster.
+The existing Alertmanager cluster will aggregate both user workload alerts as well as cluster alerts. Tenancy will be achieved by forcing specific labels onto metrics and alerts by which the central Alertmanager config can route similarly to how tenancy is achieved at the query level (see below).
+
+Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.
+Tenancy is enforced at the prometheus query layer. This is achived by leveraging the existing topology using kube-rbac-proxy [1] and prom-label-proxy [2].
+
+Differences between cluster- and user-level is mainly upgrade guarantees. The cluster-level Prometheus stack is an integral part of OpenShift and its reliability is important, keeping the stack immutable and fully testable by us allows us to rely on it to a higher degree. As we have no such possibility for user defined configurations and scrape targets, we can't provide the same guarantees.
+
+There are edge cases where an automated upgrade is not possible in the face of user workload monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
+
+Instead, the "upgradable" condition in the ClusterOperator will be set to false.
```

Comment:
@s-urbaniak @LiliC I was talking with @jwforres about this and there are two concerns:

1) this means that a user workload can effectively block upgrades (or at least the successful completion of upgrades) on the cluster.  That's not something we can live with
2) how would we know before the upgrade, that the query was not going to be valid in the new version of prometheus?  At best i'd expect us to upgrade to the new version and then find out things were broken.

I think instead of blocking the upgrade, we should simply send out alerts if there are invalid queries and let someone fix them, unless the invalid queries have a negative impact on prometheus itself?

Reply:
I'll answer in reverse:

2) We only know this because we're deeply enough involved into Prometheus to know these, which doesn't necessarily mean we'd actually catch these occurrences. Which leads me to say about 1) apart from the obvious major version change we've never had a PromQL change like this, and the worst that would happen is that the Prometheus server would reject the rule file, which if the user setup the Watchdog alerting would notify them immediately. As a side note to that we're working on surfacing better when people haven't configured alertmanager routes (as well as the urgency of this).

Reply:
Indeed, we are talking about rare edge cases which nevertheless have to be called out here. I think it is worth mentioning the concrete edge case where Prometheus might refuse to start due to rendering a dysfunctional configuration. In the case of in-cluster-monitoring we can catch this during the development cycle and e2e test failures and fix the offending ServiceMonitors. In the case of user-workload-monitoring we have much less to no control on what ServiceMonitors are being reconciled. (This is also an additional reason why those stacks are strictly separated).

Having said that we might think of better error handling in the concrete edge case where Prometheus refuses to start due to wrong configuration rendering, i.e. skipping the sections which drew it malfunctional and issue an alert here but this kind of error handling needs further investigation.

Reply:
Added the above idea of validation logic as a GA-criterion in https://issues.redhat.com/browse/MON-845